Adding concrete deployment for Concourse servers

blarghmatey opened the pull request:
We are moving to use Concourse as our continuous delivery/deployment system and this adds the code needed to build images for deploying Concourse servers in production-like environments.

blarghmatey committed:
Adding concrete configs for services to deploy Concourse

- Started adding specific configuration values
- Started adding SOPS for managing secrets during build time

blarghmatey committed:
Added configuration for Vault agent to manage secrets

- Adding configuration to allow for vault agent to manage templated secrets
- Cleaning up model values to work as a full build
- Updated Caddy to use environment var for domain setting

blarghmatey committed:
Fixing management of configuration across multiple Hashicorp products

- Adding in configuration for Consul agents
- Fixed issue with temporary file reuse over multiple calls to configure hashicorp services

shaidar commented on src/bilder/components/concourse/models.py:
```diff
@@ -206,7 +205,8 @@ class ConcourseWorkerConfig(ConcourseBaseConfig):
     )
-    tsa_public_key: Path = Field(
+    tsa_public_key: Optional[str]
+    tsa_public_key_path: Path = Field(
```
This looks a bit odd with double `Path` but saw the docs and should be fine

shaidar commented on src/bilder/components/concourse/steps.py:
```diff
@@ -75,16 +75,6 @@ def install_concourse(concourse_config: ConcourseBaseConfig, state=None, host=No
         state=state,
         host=host,
     )
-    # create configuration directory
```
Why is this no longer needed?

blarghmatey replied:
It was never actually used so I removed it to avoid confusion.

shaidar commented on src/bilder/components/hashicorp/consul/models/consul_template.py:
```diff
@@ -1,10 +1,81 @@
 
 class ConsulTemplateConfig(HashicorpConfig):
+    template: List[ConsulTemplateTemplate] = [
+        ConsulTemplateTemplate(destination=Path("/tmp/test.txt"))  # noqa: S108
```
Can we use something more descriptive for the file name?

blarghmatey replied:
That's just a placeholder for being able to populate a default configuration. It won't actually be used in real use and the destination will be overridden.

shaidar commented on src/bilder/deployments/concourse/deploy.py:
```diff
@@ -11,13 +19,150 @@
     install_concourse,
     register_concourse_service,
 )
+from bilder.components.hashicorp.consul.models.consul import Consul, ConsulConfig
```
Why is this path kinda redundant `consul.models.consul`?

blarghmatey replied:
I had started with having all of the models in a models.py file, but because I'm treating the various products in the Consul umbrella as the same module I split `models` into an enclosing module space. Perhaps we should treat each of the different services as a separate product and give them their own models.py...

shaidar commented on src/bilder/deployments/concourse/deploy.py:
```diff
@@ -11,13 +19,150 @@
+concourse_config_map = {
+    "web": partial(  # noqa: S106
+        ConcourseWebConfig,
+        admin_user="oldevops",
```
Not sure what this is insinuating  :-) First read it as old devops. I've seen us use `ol` in places and `odl` in others. Would be good to settle on one or the other.

blarghmatey replied:
I've been trying to standardize on `ol` now that the `Open Learning` designation seems to be sticking and `odl` has largely been deprecated.

shaidar commented on src/bilder/deployments/concourse/deploy.py:
```diff
@@ -11,13 +19,150 @@
+    caddyfile=Path(__file__)
+    .parent.resolve()
+    .joinpath("templates", "concourse_caddyfile.j2"),
+    plugins=[CaddyPlugin(repository="github.com/caddy-dns/route53", version="v1.1.1")],
```
Does repository need a defined scheme?

blarghmatey replied:
Most likely, yes, but punting on that for the time being in the interest of speed.

shaidar commented on src/bilder/deployments/concourse/deploy.py:
```diff
@@ -11,13 +19,150 @@
+)
+caddy_config.template_context = caddy_config.dict()
+install_caddy(caddy_config)
+caddy_config_changed = configure_caddy(caddy_config)
```
Can we come up with a more standard naming convention given that for concourse we assigned install and config concourse `install_changed` and `config_changed` and here we didn't need to assign a variable to install_caddy but then used a `caddy_config_changed` for configure caddy

blarghmatey replied:
I'll give it a shot.

blarghmatey committed:
Adding Consul service definition for Concourse web

blarghmatey committed:
Adding proxy setup for Consul DNS and Packer config

shaidar commented on .sops.yaml:
```diff
@@ -0,0 +1,12 @@
+  - path_regex: (production.env$|.*\.?production\.ya?ml)
+    kms: 'arn:aws:kms:us-east-1:610119931565:alias/infrastructure-secrets-production'
+    pgp: dfd1205a98269f34c43c9221d6260c616bd0fa33,B905F0256A801D3F1D15B5126A1D27A4BE75B1C2,51BB820DC5F14FE9
+  - kms: 'arn:aws:kms:us-east-1:610119931565:alias/infrastructure-secrets-production'
```
Duplicate to line 9?

blarghmatey replied:
No, because that is applying to any other file that doesn't match the patterns specified in the other list items.

shaidar commented on secrets/consul.env:
```diff
@@ -0,0 +1,20 @@
+#ENC[AES256_GCM,data:0BaFmmGQZYEbcf8AH8msfw58gVjZDZXDNdpaYow7wfssH4ibcWEBTkIy/DnRIekWXm3qYn1T7YPqcSHM,iv:7XeqQ0DmU+WFJlm3tCLXrWhkeIISMunO+R3hT7VvzcM=,tag:oOs2BmfVlraoccud5dx3uA==,type:comment]
+CONSUL_ENCRYPT=ENC[AES256_GCM,data:5e4nDumoNlBDt4W4oQsBMyWSG6L6ox1N,iv:rCt/hafismE3hs8givjSjLiqfM+WbcXa9kj5cuBje2w=,tag:3P6w9Jl3/mgQRwPG4Dp0bQ==,type:str]
+sops_kms__list_0__map_created_at=2021-04-22T20:55:24Z
+sops_kms__list_0__map_aws_profile=
```
Is this supposed to not have a value?

blarghmatey replied:
That's something that Sops manages automatically and not under our control.

shaidar commented on secrets/consul.env:
```diff
@@ -0,0 +1,20 @@
+sops_unencrypted_suffix=_unencrypted
+sops_kms__list_0__map_enc=AQICAHgQalNS7T35ZlcFdhF0QuKeiJAbXMUbm01pjGwHEsjRCgFn1142Wsgrc01558WjWbdXAAAAfjB8BgkqhkiG9w0BBwagbzBtAgEAMGgGCSqGSIb3DQEHATAeBglghkgBZQMEAS4wEQQMW4SyhISXXfFKYpJkAgEQgDv/pOFZ1WDui5FccZkt5t/lKq8PlMezacQaYhfRhzVVzBfCzr4vqESrSOymkPwRhSSv3c0/8BOgpav+NQ==
+sops_version=3.7.1
+sops_lastmodified=2021-04-22T20:56:00Z
```
nit-pick: Should this be `sops_last_modified`?

blarghmatey replied:
That's not something that we control, that's metadata that Sops adds automatically.

shaidar commented on src/bilder/components/hashicorp/steps.py:
```diff
@@ -130,9 +130,13 @@ def register_services(
 @deploy("Configure Hashicorp Products")
 def configure_hashicorp_product(product: HashicorpProduct, state=None, host=None):
     put_results = []
+    counter = 0
```
What's counter used for?

blarghmatey committed:
Add template for authorized worker keys and remove loop counter

- Remove a loop counter that was added while debugging but not necessary for functionality
- Adding template to populate authorized worker keys file on web nodes

shaidar commented on src/bilder/deployments/concourse/deploy.py:
```diff
@@ -11,13 +19,189 @@
-concourse_config = ConcourseBaseConfig()
-web_config = ConcourseWebConfig()
-worker_config = ConcourseWorkerConfig()
+CONCOURSE_WEB_HOST_COMMUNICATION_PORT = 2222
```
Should this be in magic numbers?

blarghmatey replied:
I opted to keep it here since it's only relevant to Concourse and not a service that we're likely to use in multiple places as compared to DNS, Postgres, etc.

shaidar commented on src/bilder/deployments/concourse/deploy.py:
```diff
@@ -11,13 +19,189 @@
+                name="concourse",
+                port=CONCOURSE_WEB_HOST_COMMUNICATION_PORT,
+                tags=[CONCOURSE_WEB_NODE_TYPE],
+                check=ConsulServiceTCPCheck(tcp="localhost:2222"),
```
Maybe we should use `CONCOURSE_WEB_HOST_COMMUNICATION_PORT` here instead of statically defining it?

shaidar commented on src/bilder/deployments/concourse/deploy.py:
```diff
@@ -11,13 +19,189 @@
+    Vault(
+        configuration=VaultAgentConfig(
+            vault=VaultConnectionConfig(
+                address="https://active.vault.service.consul:8200",
```
Vault isn't in magic numbers? Should it be?

blarghmatey committed:
Reusing Concourse port constant

shaidar commented on src/bilder/deployments/concourse/deploy.py:
```diff
@@ -11,13 +19,189 @@
+            ],
+        )
+    ),
+    Consul(version="1.9.5", configuration=consul_configuration),
```
Should version numbers (this and a few others) maybe be defined at the beginning of the file so that it's easier to see what versions are being deployed and/or changed?

blarghmatey committed:
Adding Vault HTTP to magic numbers

blarghmatey committed:
Consolidating version specifiers

shaidar commented:
👍
